### PR TITLE
Ben/assets rpc

### PR DESCRIPTION
**Pull request by benluelo**

### CODE REVIEW

**Review by hussein-aitlahcen — COMMENTED** [review]
Any reason for `assets` pallet to be moved under a subdirectory?

File: frame/assets/rpc/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+impl<C, M> Assets<C, M> {
+	/// Create new `SumStorage` instance with the given reference to the client.
+	pub fn new(client: Arc<C>) -> Self {
+		Self { client, _marker: Default::default() }
+	}
+}
+
+/// Error type of this RPC api.
```
Comment (dzmitry-lahoda) [inline frame/assets/rpc/src/lib.rs:36]: please comment when going uncomment this code?

File: frame/assets/rpc/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+		runtime_api_result.map_err(|e| {
+			RpcError {
+				code: ErrorCode::ServerError(9876), // No real reason for this value
+				message: "Something wrong".into(),
+				data: Some(format!("{:?}", e).into()),
+			}
+		})
+	}
+
+	// fn get_sum(&self, at: Option<<Block as BlockT>::Hash>) -> Result<u32> {
```
Comment (dzmitry-lahoda) [inline frame/assets/rpc/src/lib.rs:89]: please comment when going uncomment this code?

**Comment by dzmitry-lahoda** [comment]
> Any reason for `assets` pallet to be moved under a subdirectory?

@benluelo afaik you can do root folder with pallet, and then do 2 child subs with cargo tomls for apis. i have seen this. its ok. so we will not violated rule that fram/pallets folders contains pallets in direct children, not subchildern

File: frame/assets/rpc/src/lib.rs (hunk 2 above)
Comment (benluelo) [reply frame/assets/rpc/src/lib.rs:89]: removed

File: frame/assets/rpc/src/lib.rs (hunk 1 above)
Comment (benluelo) [reply frame/assets/rpc/src/lib.rs:36]: removed

**Comment by benluelo** [comment]
> Any reason for `assets` pallet to be moved under a subdirectory?

It was just to put it beside the rpc and runtime-api. I've moved it back to `frame/assets`

**Comment by benluelo** [comment]
The integration tests for the rpcs aren't implemented as of yet (just stubs with a failing line), I'm not sure what to test.

**Comment by hussein-aitlahcen** [comment]
The need to be rebased I think

**Review by dzmitry-lahoda — DISMISSED** [review]

**Comment by hussein-aitlahcen** [comment]
We probably need to merge/rebase on main

File: docs/rpc.md [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+authors = ["Composable Developers"]
+homepage = "https://composable.finance"
+edition = "2021"
+rust-version = "1.56"
+
+[package.metadata.docs.rs]
+targets = ["x86_64-unknown-linux-gnu"]
+
+[dependencies]
+sp-api = { default-features = false, git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.13" }
```
Comment (andor0) [inline docs/rpc.md:45]: We already use `branch = "polkadot-v0.9.16"`.

File: frame/assets/rpc/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+homepage = "https://composable.finance"
+edition = "2021"
+rust-version = "1.56"
+
+[package.metadata.docs.rs]
+targets = ["x86_64-unknown-linux-gnu"]
+
+[dependencies]
+# substrate primitives
+sp-api = { default-features = false, git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.13" }
```
Comment (andor0) [inline frame/assets/rpc/Cargo.toml:14]: `polkadot-v0.9.16`

File: integration-tests/runtime-tests/src/types/interfaces/assets/definitions.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+          isOptional: true,
+        },
+      ],
+      type: "AssetsBalance"
+    },
+  },
+  types: {
+    CurrencyId: "u128",
+    AssetsBalance: "u128",
+    // Balance: "u128",
```
Comment (andor0) [inline integration-tests/runtime-tests/src/types/interfaces/assets/definitions.ts:26]: can remove it?

Comment (benluelo) [reply integration-tests/runtime-tests/src/types/interfaces/assets/definitions.ts:26]: Remove what?

Comment (andor0) [reply integration-tests/runtime-tests/src/types/interfaces/assets/definitions.ts:26]:
```
 // Balance: "u128",
 ```

**Comment by benluelo** [comment]
Still working on a fix for the broken CI.

**Comment by benluelo** [comment]
Not sure why rustfmt is failing, it was failing locally as well

File: Cargo.toml [hunk 6]
```diff
@@ -13,28 +13,34 @@ edition = "2018"
...
 [workspace]
 exclude = ["frame/transaction-fee"]
 members = [
-    "runtime/common",
-    "runtime/primitives",
     "frame/*",
+    "frame/assets/rpc",
+    "frame/assets/runtime-api",
+    "frame/crowdloan-rewards/rpc",
+    "frame/crowdloan-rewards/runtime-api",
```
Comment (hussein-aitlahcen) [inline Cargo.toml:30]: This isn't working if we just set the `*`?

File: frame/composable-support/src/lib.rs [hunk 7]
```diff
@@ -14,3 +14,6 @@
 
 pub mod collections;
 pub mod validation;
+
+// #[cfg(feature = "std")]
```
Comment (hussein-aitlahcen) [inline frame/composable-support/src/lib.rs:18]: Remove this :)

File: Cargo.toml (hunk 6 above)
Comment (benluelo) [reply Cargo.toml]: Seems like it does, I thought that `frame/*` would only match the first level of nested directories. But it works so I removed them :+1:

**Review by hussein-aitlahcen — APPROVED** [review]
Well done :)

File: frame/assets/Cargo.toml [hunk 8]
```diff
@@ -14,26 +14,26 @@ frame-benchmarking = { default-features = false, optional = true, git = 'https:/
 frame-support = { default-features = false, git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.16" }
 frame-system = { default-features = false, git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.16" }
 
+sp-api = { default-features = false, git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.16" }
+# sp-blockchain = { default-features = false, git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.16" }
```
Comment (andor0) [inline frame/assets/Cargo.toml:18]: can remove it?

File: integration-tests/runtime-tests/src/types/interfaces/definitions.ts [hunk 9]
```diff
@@ -1,71 +1,4 @@
...
-    CumulusPrimitivesParachainInherentParachainInherentData: 'ParachainInherentData',
-    DaliRuntimeOriginCaller: "Null",
-    ComposableTraitsVaultVaultConfig: "Null",
-    XcmVersionedMultiAsset: "Null",
-  }
-};
+export { default as crowdloanRewards } from "./crowdloanRewards/definitions";
+export { default as assets } from "./assets/definitions";
+// export { default as composableTraits } from "./composableTraits/definitions";
+// export { default as dutchAuction } from "./dutchAuction/definitions";
```
Comment (andor0) [inline integration-tests/runtime-tests/src/types/interfaces/definitions.ts:4]: can remove it?

File: integration-tests/runtime-tests/src/utils/testSetup.ts [hunk 10]
```diff
@@ -13,19 +13,31 @@ global.testSudoCommands = true;
...
 export async function runBefore() {
-  // extract all types from definitions - fast and dirty approach, flatted on 'types'
-  const types = Object.values(definitions).reduce((res, {types}): object => ({...res, ...types}), {});
+  const chai = require('chai');
+  const BN = require('bn.js');
+
+  // Enable and inject BN dependency
+  chai.use(require('chai-bn')(BN));
+
+  // console.log(JSON.stringify(definitions, null, 4));
```
Comment (andor0) [inline integration-tests/runtime-tests/src/utils/testSetup.ts:22]: can remove it?

File: integration-tests/runtime-tests/src/utils/testSetup.ts [hunk 11]
```diff
@@ -13,19 +13,31 @@ global.testSudoCommands = true;
...
+  const rpc = Object.keys(definitions).reduce((accumulator, key) => ({ ...accumulator, [key]: definitions[key].rpc }), {});
+  const types = Object.values(definitions).reduce((accumulator, { types }) => ({ ...accumulator, ...types }), {});
+  // console.log("rpc: ", JSON.stringify(rpc, null, 4));
+  // console.log("types: ", JSON.stringify(types, null, 4));
```
Comment (andor0) [inline integration-tests/runtime-tests/src/utils/testSetup.ts:26]: can remove it?

**Review by andor0 — APPROVED** [review]

**Review by dzmitry-lahoda — APPROVED** [review]

**Pull request merged**